### PR TITLE
UF-7785: Added logic and integration towards document service

### PR DESCRIPTION
**Pull request by josc04**
- Added document service integration
- Added logic in workers to read and update document inrformation
- Added more tasks/workers for better separation of concerns

### CODE REVIEW

File: src/test/java/se/sundsvall/esigning/integration/document/model/DocumentMultipartFileTest.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import generated.se.sundsvall.comfactfacade.Document;
+import wiremock.org.apache.commons.io.FileUtils;
+
+@ExtendWith(MockitoExtension.class)
+class DocumentMultipartFileTest {
+
+	@Mock
+	File fileMock;
```
Comment (foobar-1337) [inline src/test/java/se/sundsvall/esigning/integration/document/model/DocumentMultipartFileTest.java:24]: private?

File: src/main/java/se/sundsvall/esigning/integration/document/DocumentClient.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+	Document getDocument(@PathVariable("registrationNumber") String registrationNumber);
+
+	@PatchMapping(path = "documents/{registrationNumber}")
+	Document updateDocument(@PathVariable("registrationNumber") String registrationNumber, DocumentUpdateRequest body);
+
+	@GetMapping(path = "documents/{registrationNumber}/files/{documentDataId}")
+	ResponseEntity<ByteArrayResource> getDocumentData(@PathVariable("registrationNumber") String registrationNumber, @PathVariable("documentDataId") String documentDataId);
+
+	@PutMapping(path = "/documents/{registrationNumber}/files", consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
+	Void addFileToDocument(@PathVariable("registrationNumber") String registrationNumber, @RequestPart("document") String document, @RequestPart("documentFile") MultipartFile documentFile);
```
Comment (foobar-1337) [inline src/main/java/se/sundsvall/esigning/integration/document/DocumentClient.java:33]: Why is document a String?
Isn't DocumentDataCreateRequest `the correct type for the PUT-operation?

Comment (josc04) [reply src/main/java/se/sundsvall/esigning/integration/document/DocumentClient.java:33]: According to the API the requestpart for the DocumentCreateRequest is to be sent as a String:
```
	@PutMapping(path = "/{registrationNumber}/files", consumes = {MULTIPART_FORM_DATA_VALUE}, produces = {APPLICATION_JSON_VALUE, APPLICATION_PROBLEM_JSON_VALUE})
	@Operation(summary = "Add document file data (or replace existing if filename already exists on the document object)")
	@ApiResponse(responseCode = "200", description = "Successful operation", useReturnTypeSchema = true)
	@ApiResponse(responseCode = "404", description = "Not found", content = @Content(mediaType = APPLICATION_PROBLEM_JSON_VALUE, schema = @Schema(implementation = Problem.class)))
	public ResponseEntity<Void> addOrReplaceFile(
		@Parameter(name = "registrationNumber", description = "Document registration number", example = "2023-2281-1337") @PathVariable("registrationNumber") String registrationNumber,
		@RequestPart("document") @Schema(description = "Document", implementation = DocumentDataCreateRequest.class) String documentDataString,
		@RequestPart(value = "documentFile") MultipartFile documentFile) throws JsonProcessingException {
```
I tried to send it as an json object but the service responded with 400, and that requestPart 'document' was not present in the request on such requests.

I'm not sure, but I guess that there is limitations when using requestPart annotations in the API?

Comment (foobar-1337) [reply src/main/java/se/sundsvall/esigning/integration/document/DocumentClient.java:33]: image

According to the API, "document" should be a DocumentDataCreateRequest

Comment (josc04) [reply src/main/java/se/sundsvall/esigning/integration/document/DocumentClient.java:33]: Yes but it doesn't work if the object is sent as json in the requestPart. As I said, there are limitations when using requestPart - one is that you can not send json as a requestPart. It has to be converted to string to work (I have googled it and the suggested solution is to serialize the object to a string).

Comment (foobar-1337) [reply src/main/java/se/sundsvall/esigning/integration/document/DocumentClient.java:33]: I dont understand?

Isn't json sent as string data?
Isn't feign serializing the Java object into a string? (what is sent if it's not a string with json-data?)

**Comment by foobar-1337** [comment]
approved

**Pull request merged**